infer/src/integration/JsonReports.ml

```diff
@@ -140,6 +178,80 @@ end) : Printer with type elt = P.elt = struct
         ()
 end
 
+module type PrinterSarif = sig
+  type elt
+
+  val pp_open : F.formatter -> unit -> unit
+
+  val pp_results : F.formatter -> unit -> unit
+
+  val pp_close : F.formatter -> unit -> unit
+
+  val pp_rules : F.formatter -> elt -> unit
+
+  val pp : F.formatter -> elt -> unit
+end
+
+module MakeSarifListPrinter (P : sig
+  type elt
+
+  val rules_to_string : elt -> string option
+
+  val to_string : elt -> string option
+end) : PrinterSarif with type elt = P.elt = struct
+  include P
+
+  let is_first_item = ref true
+  
+  let pp_n_spaces n fmt =
+    for _ = 1 to n do
+      F.pp_print_char fmt ' '
+    done
+
+  let pp_open fmt () =
+    is_first_item := true ;
+    F.fprintf fmt "{@\n" ;
+    F.fprintf fmt "%t%s: %s,@\n" (pp_n_spaces 2) "\"schema\"" "\"http://json.schemastore.org/sarif-2.1.0\"" ;
+    F.fprintf fmt "%t%s: %s,@\n" (pp_n_spaces 2) "\"version\"" "\"2.1.0\"" ;
+    F.fprintf fmt "%t%s: [@\n"  (pp_n_spaces 2) "\"runs\"" ;
+    F.fprintf fmt "%t{@\n"  (pp_n_spaces 4) ;
+    F.fprintf fmt "%t%s: {@\n"  (pp_n_spaces 6) "\"tool\"" ;
+    F.fprintf fmt "%t%s: {@\n"  (pp_n_spaces 8) "\"driver\"" ;
+    F.fprintf fmt "%t%s: %s,@\n"  (pp_n_spaces 10) "\"name\"" "\"Infer\"" ;
+    F.fprintf fmt "%t%s: %s,@\n"  (pp_n_spaces 10) "\"informationUri\"" "\"https://github.com/facebook/infer\"" ;
+    F.fprintf fmt "%t%s: \"%d.%d.%d\",@\n"  (pp_n_spaces 10) "\"version\"" Version.major Version.minor Version.patch ;
+    F.fprintf fmt "%t%s: [@?"  (pp_n_spaces 10) "\"rules\""
+
+  let pp_results fmt () =
+    is_first_item := true ;
+    F.fprintf fmt "]@\n" ;
+    F.fprintf fmt "%t%s@\n"  (pp_n_spaces 8) "}" ;
+    F.fprintf fmt "%t%s,@\n"  (pp_n_spaces 6) "}" ;
+    F.fprintf fmt "%t%s: [@?"  (pp_n_spaces 6) "\"results\""
+
+  let pp_close fmt () = 
+    F.fprintf fmt "]@\n" ;
+    F.fprintf fmt "%t%s@\n" (pp_n_spaces 4) "}" ;
+    F.fprintf fmt "%t%s@\n" (pp_n_spaces 2) "]" ;
+    F.fprintf fmt "}@\n@?"
```

Comment:
Instead of formatting json by hand here, would it make sense to add these types to the .atd file and use the auto-generated code to print these fields?